allennlp/common/file_utils.py

```diff
@@ -161,29 +172,40 @@ def cached_path(
 
     elif parsed.scheme == "":
         # File, but it doesn't exist.
-        raise FileNotFoundError("file {} not found".format(url_or_filename))
+        raise FileNotFoundError(f"file {url_or_filename} not found")
 
     else:
         # Something unknown
-        raise ValueError("unable to parse {} as a URL or as a local path".format(url_or_filename))
+        raise ValueError(f"unable to parse {url_or_filename} as a URL or as a local path")
 
     if extraction_path is not None:
-        # No need to extract again.
+        # If the extracted directory already exists (and is non-empty), then no
+        # need to extract again unless `force_extract=True`.
         if os.path.isdir(extraction_path) and os.listdir(extraction_path) and not force_extract:
             return extraction_path
 
         # Extract it.
         with FileLock(file_path + ".lock"):
             shutil.rmtree(extraction_path, ignore_errors=True)
-            os.makedirs(extraction_path)
-            if is_zipfile(file_path):
-                with ZipFile(file_path, "r") as zip_file:
-                    zip_file.extractall(extraction_path)
-                    zip_file.close()
-            else:
-                tar_file = tarfile.open(file_path)
-                tar_file.extractall(extraction_path)
-                tar_file.close()
+
+            # We extract first to a temporary directory in case something goes wrong
+            # during the extraction process so we don't end up with a corrupted cache.
+            tmp_extraction_dir = tempfile.mkdtemp(dir=os.path.split(extraction_path)[0])
+            try:
+                if is_zipfile(file_path):
+                    with ZipFile(file_path, "r") as zip_file:
+                        zip_file.extractall(tmp_extraction_dir)
+                        zip_file.close()
+                else:
+                    tar_file = tarfile.open(file_path)
+                    tar_file.extractall(tmp_extraction_dir)
+                    tar_file.close()
+                # Extraction was successful, rename temp directory to final
+                # cache directory.
+                os.makedirs(extraction_path)
+                os.replace(tmp_extraction_dir, extraction_path)
```

Comment:
One more comment: Doesn't the first line create all the directories, including the target one? If we fail between these two lines, I think it'll conclude that the archive has been extracted and won't do it again?

In summary, I think that first one has to be `os.makedirs(os.path.split(extraction_path)[0])`.

Reply:
Ah yes, good point.

Reply:
Just fixed. After taking another look, I realized we can just remove that first line altogether because at that point the base directory should already exist.